apis/Google.Cloud.Firestore/Google.Cloud.Firestore/AggregateQuery.cs

```diff
@@ -0,0 +1,129 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and 
+// limitations under the License.
+
+using Google.Api.Gax;
+using Google.Api.Gax.Grpc;
+using Google.Cloud.Firestore.V1;
+using Google.Protobuf;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using static Google.Cloud.Firestore.V1.StructuredAggregationQuery.Types;
+
+namespace Google.Cloud.Firestore;
+
+/// <summary>
+/// A query for running an aggregation over a [StructuredQuery][google.firestore.v1.StructuredQuery]. Currently only "count(*)" aggregation is supported.
+/// </summary>
+public sealed class AggregateQuery : IEquatable<AggregateQuery>
+{
+    private readonly Query _query;
+
+    /// <summary>
+    /// The "alias" to specify in the <see cref="RunAggregationQueryRequest"/> proto when running a count query.
+    /// The actual value is not meaningful, but will be used to get the count out of the <see cref="RunAggregationQueryResponse"/>.
+    /// </summary>
+    private const string CountAlias = "Count";
+
+    private readonly IReadOnlyList<Aggregation> _aggregations;
+
+    internal AggregateQuery(Query query) =>
+        _query = GaxPreconditions.CheckNotNull(query, nameof(query));
+
+    private AggregateQuery(Query query, IReadOnlyList<Aggregation> aggregations)
+    {
+        _query = GaxPreconditions.CheckNotNull(query, nameof(query));
+        _aggregations = aggregations;
+    }
+
+    internal AggregateQuery WithAggregation(Aggregation aggregation)
+    {
+        if (aggregation.Count != null)
+        {
+            aggregation.Alias = CountAlias;
+        }
+        var newAggregations = _aggregations == null ? new List<Aggregation>() : new List<Aggregation>(_aggregations);
+        newAggregations.Add(aggregation);
+        return new AggregateQuery(_query, newAggregations);
+    }
+
+    /// <summary>
+    /// Asynchronously takes a snapshot of the result after applying the aggregate functions on the query.
+    /// </summary>
+    /// <param name="cancellationToken">A cancellation token for the operation.</param>
+    /// <returns>A <see cref="AggregateQuerySnapshot"/> which contains results of the Aggregate functions.</returns>
+    public Task<AggregateQuerySnapshot> GetSnapshotAsync(CancellationToken cancellationToken = default) =>
+        GetSnapshotAsync(null, cancellationToken);
+
+    internal async Task<AggregateQuerySnapshot> GetSnapshotAsync(ByteString transactionId, CancellationToken cancellationToken)
+    {
+        IAsyncEnumerable<RunAggregationQueryResponse> responseStream = GetAggregationQueryResponseStreamAsync(transactionId, cancellationToken);
+        Timestamp? readTime = null;
+        Value count = null;
+        await responseStream.ForEachAsync(response =>

```

Comment:
It might be worth extracting this lambda expression into a local method (e.g. `ProcessResponse`) for readability - I probably didn't do it when writing the Firestore client initially as they didn't exist back then ;)